### PR TITLE
Ui tweaks

### PR DESCRIPTION
**Pull request by jytang**
transition ui
leaderboard ui tweaks
gold ui tweaks
main menu ｓｋえｌえｔおｎ

### CODE REVIEW

**Review by AznCelebration — APPROVED** [review]

File: src/game/game_state.cpp [hunk 1]
```diff
@@ -105,10 +105,14 @@ void GameState::set_phase(Phase* phase) {
         curr_phase->s_setup();
     }
     else {
-        if (curr_phase)
+        if (curr_phase == nullptr) {
```
Comment (metakirby5) [inline src/game/game_state.cpp]: ...but why?

File: src/ui/score_ui.cpp [hunk 2]
```diff
@@ -63,13 +63,33 @@ ScoreUI::ScoreUI(float aspect)
...
 
     events::ui::disable_scoreboard.connect([&]() {
+        disable_animated();
+    });
+}
+
+void ScoreUI::enable_animated() {
+    enable();
+
+    for (int i = 0; i < entries.size(); ++i) {
```
Comment (metakirby5) [inline src/ui/score_ui.cpp]: Enhanced for loop please

File: src/ui/score_ui.cpp [hunk 3]
```diff
@@ -63,13 +63,33 @@ ScoreUI::ScoreUI(float aspect)
...
+        entries[i]->animate_alpha(1.f, 0.5f);
+    }
+
+    bg.animate_alpha(0.7f, 0.5f);
+}
+
+void ScoreUI::disable_animated() {
+    for (int i = 0; i < entries.size(); ++i) {
```
Comment (metakirby5) [inline src/ui/score_ui.cpp]: E N H A N C E

File: src/ui/score_ui.cpp [hunk 4]
```diff
@@ -63,13 +63,33 @@ ScoreUI::ScoreUI(float aspect)
...
         // Do gold delta sequence.
         animate_deltas();
     });
 
     events::ui::disable_scoreboard.connect([&]() {
+        disable_animated();
+    });
+}
+
+void ScoreUI::enable_animated() {
```
Comment (metakirby5) [inline src/ui/score_ui.cpp:77]: Can't we have the base `UI` class just animate all its children?

File: src/ui/transition_ui.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    events::ui::transition_fade.connect([&](float seconds, std::function<void()> do_apex) {
+        fade(seconds,
+             Color::BLACK,
+             do_apex);
+    });
+}
+
+void TransitionUI::fade(float seconds, Color c, std::function<void()> do_apex) {
+    //bg.set_color(c);
```
Comment (metakirby5) [inline src/ui/transition_ui.cpp]: delet

File: src/ui/transition_ui.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    });
+
+    events::ui::transition_fade.connect([&](float seconds, std::function<void()> do_apex) {
+        fade(seconds,
+             Color::BLACK,
+             do_apex);
+    });
+}
+
+void TransitionUI::fade(float seconds, Color c, std::function<void()> do_apex) {
```
Comment (metakirby5) [inline src/ui/transition_ui.cpp]: Could really use some comments here

File: src/ui/transition_ui.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void TransitionUI::wipe_reset() {
+    disable();
+    bg.set_start_x(initial_start_x);
+    bg.set_start_y(initial_start_y);
+}
+
+void TransitionUI::wipe(float seconds, Color c, Direction dir, std::function<void()> do_apex) {
+    //bg.set_color(c);
```
Comment (metakirby5) [inline src/ui/transition_ui.cpp]: delet

File: src/ui/transition_ui.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            break;
+        case Direction::RIGHT:
+            new_start_x = -125.f * aspect;
+            target_x = 100.f * aspect;
+            break;
+        default:
+            break;
+    }
+
+    // Set starting position of wipe.
```
Comment (metakirby5) [inline src/ui/transition_ui.cpp:92]: Consider parameterizing `wipe_reset` and using it here.

File: src/ui/transition_ui.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    bg.set_start_x(new_start_x);
+    bg.set_start_y(new_start_y);
+
+    // Animate to cover the screen.
+    bg.animate_to(initial_start_x, initial_start_y, seconds / 3.f, [&, do_apex, seconds, target_x, target_y]() {
+        do_apex();
+        Timer::get()->do_after(std::chrono::milliseconds((int) (seconds / 3.f * 1000)), [&, seconds, target_x, target_y]() {
+            // Complete animation to target position.
+            bg.animate_to(target_x, target_y, seconds / 3.f, [&]() {
+                // 3 nested lambdas.
```
Comment (metakirby5) [inline src/ui/transition_ui.cpp:102]: i know that feel

File: vs/vs.vcxproj [hunk 10]
```diff
@@ -130,7 +130,7 @@ xcopy /F /Y "$(SolutionDir)\packages\boost_signals-vc140.1.64.0.0\lib\native\add
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>TurnOffAllWarnings</WarningLevel>
```
Comment (metakirby5) [inline vs/vs.vcxproj:133]: 🤔

**Review by metakirby5 — CHANGES_REQUESTED** [review]

File: src/ui/score_ui.cpp (hunk 4 above)
Comment (jytang) [reply src/ui/score_ui.cpp:77]: well different UIs animate in different ways. the base UI does have a default thingy though.

File: src/ui/transition_ui.cpp (hunk 6 above)
Comment (jytang) [reply src/ui/transition_ui.cpp]: NO

**Pull request merged**